### ISSUE TITLE
Lights randomly come back on after shut off

### ISSUE THREAD

**Issue by broyuken** [basnijholt/adaptive-lighting#10]
This has only happened to me twice, and I don't have a way to replicate it, but I will turn off a light using lovelace and after 10-15 seconds or so it turns itself back on. If I find a way to replicate it I can turn on debugging but I have no way to track it down so it wouldn't do much good right now. I am running d67479d now, and if I see it happen with this version I will let you know but like I said this has only happened twice over the week or so I've been running AL.

**Comment by basnijholt** [comment]
Would it be possible to enable debugging logging with:
```
logger:
  default: warning
  logs:
    custom_components.adaptive_lighting: debug
```
and post the log after it occurred?

I have a suspicion as to why it is happening but not sure.

**Comment by broyuken** [comment]
I just enabled debugging to see if it happens again. It just happened again on the latest version before I turned on debugging. I believe the only lights I have seen this occur on are Sengled Zigbee bulbs. Hopefully I can catch the full debug going forward.

**Comment by basnijholt** [comment]
OK great! 

I already anticipated that this might happen, you can see my comment in the code here
https://github.com/basnijholt/adaptive-lighting/blob/466c4b62f37192beab43bfb997cb3fc8c40a7401/custom_components/adaptive_lighting/switch.py#L1328-L1333

**Comment by broyuken** [comment]
This has happened a bunch of times since I last posted but unfortunately I messed up my Zha yesterday and had to turn off the debug to fix it. I’ll try to get it going again now that things seem to be stable

**Comment by broyuken** [comment]
I caught one this evening in my sons room. Turned off the light via home assistant and it turned back on a few seconds later. 

```
2020-10-18 20:19:23 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] State of 'light.justins_light' didn't change wrt change event nr. 0 (context.id=adapt_lgt_769d_interval_938)
2020-10-18 20:19:23 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Justins Bedroom: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.justins_light', 'transition': 45.0, 'brightness': 108, 'color_temp': 500} with context.id='adapt_lgt_769d_interval_938'
2020-10-18 20:19:23 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.justins_light']')' event with context.id='adapt_lgt_769d_interval_938'
2020-10-18 20:19:24 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.justins_light' 'state_changed' event: '{'min_mireds': 154, 'max_mireds': 500, 'brightness': 108, 'color_temp': 500, 'hs_color': (0, 0), 'rgb_color': (255, 255, 255), 'xy_color': (0.323, 0.329), 'off_brightness': None, 'friendly_name': 'Justins Light', 'icon': 'mdi:lamp', 'supported_features': 59}' with context.id='adapt_lgt_769d_interval_938'
2020-10-18 20:19:36 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_off('['light.justins_light']', transition=None)' event with context.id='c516bd2511a011eb9e71e3aac25616a3'
2020-10-18 20:20:09 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.justins_light' 'state_changed' event: '{'min_mireds': 154, 'max_mireds': 500, 'brightness': 108, 'color_temp': 500, 'hs_color': (0, 0), 'rgb_color': (255, 255, 255), 'xy_color': (0.323, 0.329), 'off_brightness': None, 'friendly_name': 'Justins Light', 'icon': 'mdi:lamp', 'supported_features': 59}' with context.id='d89b48aa11a011eb98e8b794d03648fc'
2020-10-18 20:20:09 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Justins Bedroom: Detected an 'off' → 'on' event for 'light.justins_light' with context.id='d89b48aa11a011eb98e8b794d03648fc'
2020-10-18 20:20:09 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Justins Bedroom: '_adapt_lights(['light.justins_light'], 1.0, force=True, context.id=adapt_lgt_769d_light_event_939)' called
2020-10-18 20:20:09 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Justins Bedroom: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.justins_light', 'transition': 1.0, 'brightness': 107, 'color_temp': 500} with context.id='adapt_lgt_769d_light_event_939'
2020-10-18 20:20:09 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.justins_light']')' event with context.id='adapt_lgt_769d_light_event_939'
2020-10-18 20:20:09 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.justins_light' 'state_changed' event: '{'min_mireds': 154, 'max_mireds': 500, 'brightness': 107, 'color_temp': 500, 'hs_color': (0, 0), 'rgb_color': (255, 255, 255), 'xy_color': (0.323, 0.329), 'off_brightness': None, 'friendly_name': 'Justins Light', 'icon': 'mdi:lamp', 'supported_features': 59}' with context.id='adapt_lgt_769d_light_event_939'
2020-10-18 20:20:17 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_off('['light.justins_light']', transition=None)' event with context.id='dd58269f11a011eba1913116e45fac2e'
2020-10-18 20:20:55 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Justins Bedroom: '_adapt_lights(['light.justins_light'], None, force=False, context.id=adapt_lgt_769d_interval_940)' called
2020-10-18 20:22:25 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Justins Bedroom: '_adapt_lights(['light.justins_light'], None, force=False, context.id=adapt_lgt_769d_interval_941)' called
2020-10-18 20:23:55 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Justins Bedroom: '_adapt_lights(['light.justins_light'], None, force=False, context.id=adapt_lgt_769d_interval_942)' called
2020-10-18 20:25:25 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Justins Bedroom: '_adapt_lights(['light.justins_light'], None, force=False, context.id=adapt_lgt_769d_interval_943)' called
2020-10-18 20:26:55 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Justins Bedroom: '_adapt_lights(['light.justins_light'], None, force=False, context.id=adapt_lgt_769d_interval_944)' called
2020-10-18 20:28:25 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Justins Bedroom: '_adapt_lights(['light.justins_light'], None, force=False, context.id=adapt_lgt_769d_interval_945)' called
2020-10-18 20:29:55 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Justins Bedroom: '_adapt_lights(['light.justins_light'], None, force=False, context.id=adapt_lgt_769d_interval_946)' called
2020-10-18 20:31:25 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Justins Bedroom: '_adapt_lights(['light.justins_light'], None, force=False, context.id=adapt_lgt_769d_interval_947)' called
```

**Comment by broyuken** [comment]
Any ideas?

**Comment by basnijholt** [comment]
I see that this happened:
- at 20:19:36 light.turn_off called
- at 20:20:09, the light reports it's on after polling, so AL proceeds to adapt it
- at 20:20:17, something/someone called light.turn_off again, I presume that's you clicking the off button?

This is not really a bug of AL, but rather that the light reports that it's on, while it is actually off. Still there might be something that we can do. Maybe I can ignore any `off` -> `on` events that do not result from `light.turn_on` for a minute or so. What do you think?

**Comment by broyuken** [comment]
that could be, my zigbee network has been a little flakey lately. Unfortunately most of my zigbee devices are sengled bulbs which are only endpoint devices, not routers. Working on trying to rectify that.

**Comment by broyuken** [comment]
I'm sure I'm not the only one with a less than 100% stable zigbee network, so if there is some way to ignore those events then it would probably be for the best. Especially if it was an optional configuration.

**Comment by basnijholt** [comment]
I switched to ZHA and I am now seeing these problems too. Unfortunately, it doesn't seem like there is a lot to do. I will probably switch back to deCONZ... :'(

My issue is reported here https://github.com/zigpy/zigpy-deconz/issues/140.

**Comment by broyuken** [comment]
I will watch that thread but I may need to pick up a new zigbee stick. What are you using with deconz?

**Comment by basnijholt** [comment]
I have all my devices listed here https://github.com/basnijholt/home-assistant-config/#my-devices

**Comment by danielbrunt57** [comment]
> 
> 
> that could be, my zigbee network has been a little flakey lately. Unfortunately most of my zigbee devices are sengled bulbs which are only endpoint devices, not routers. Working on trying to rectify that.

I tried HUSBZB and also Conbee II but had hiccups.
I'm much happier with [https://electrolama.com/projects/zig-a-zig-ah/] (it has an external antenna attached to the dongle)(url)
I'm also using Sengled in bedroom but Osram Lightify in office and currently do not have random turn-ons.  I did before with the Nortek and Conbee but with ERIA Tunable White A19 in the bedroom.  In hindsight, it might have been the zigbee controller.  I might try the Eria again...

**Comment by profesaurus** [comment]
I'm having this same issue. It started after I installed adaptive lighting. My Sengled bulbs will get turned off by an automation and then just turn back on for no reason. It seems to only happen at night when adaptive lighting is running.

**Comment by basnijholt** [comment]
@profesaurus, could you please do what I wrote here https://github.com/basnijholt/adaptive-lighting#having-problems ?

**Comment by danielbrunt57** [comment]
> I'm having this same issue. It started after I installed adaptive lighting. My Sengled bulbs will get turned off by an automation and then just turn back on for no reason. It seems to only happen at night when adaptive lighting is running.

Is the 'off automation' an automation in HA?
Can you verify that after the off automation runs that their status in HA shows `off`?
Also, after they've randomly turned on, does their status in HA show `on`?

**Comment by profesaurus** [comment]
@basnijholt I have adjusted HA to log debug outputs from AL.  If/when it happens again I'll post here with the logs.

@danielbrunt57

The automation is an automation in HA.  This is the only action in the automation:

![image](https://user-images.githubusercontent.com/6362518/101072578-196c0780-3553-11eb-9d0f-267d69313807.png)

Here's the logbook entry for when the automation turned the lights off, but then back on:

![image](https://user-images.githubusercontent.com/6362518/101072693-428c9800-3553-11eb-95a5-53adc1f16236.png)

The top log is interesting because the lights are turned on by the action above....which is a light.turn_off service???

Also, that automation is only triggered once at the time the lights initially turn off:

![image](https://user-images.githubusercontent.com/6362518/101073342-2806ee80-3554-11eb-8be4-7637c7581432.png)

The history does show the lights turning off, then back on:

![image](https://user-images.githubusercontent.com/6362518/101072882-84b5d980-3553-11eb-8b0e-b9389d127838.png)

I did find a [post](https://community.home-assistant.io/t/circadian-lighting-possible-fix-to-keep-light-groups-off/221188) on the HA forums about an issue with zha light groups and Circadian Lighting and AL.

I am using zha light groups, so as a test, in AL's settings I unchecked the light groups and checked the individual lights within the groups.  I don't know if this will fix it or if I'll need to try the recommended automation in the post.

Let me know if you have any other questions, I'm happy to help.

**Comment by danielbrunt57** [comment]
Very interesting.  
I have 2 Sengled bulbs (ZHA) in master bedroom.  They are motion controlled by Entity Controller, not an HA automation.  I've no issues with them turning off and staying off.  They are now in a group light.bedroom_lights.  I say now because they were in group.bedroom_lights hence the reason for the "#" in the code below as that was needed for EC to observe and ignore changes in the light entities' brightness and color attributes by CL and now AL.
```
  bedroom_lights_control:
    friendly_name: Bedroom Lights Control
    sensor: binary_sensor.lumi_sensor_motion_bedroom
    sensor_type: duration
    state_attributes_ignore:
      - color_temp
      - brightness
    entity: light.bedroom_lights
    override: 
      - input_boolean.sleep_mode
#    state_entities:
#      - light.bedroom_lamp_1_level_light_color_on_off
#      - light.bedroom_lamp_2_level_light_color_on_off
    service_data:
      brightness: 30
    service_data_on: 
      transition: 3
    service_data_off: 
      transition: 5
    start_time: sunset - 00:30:00
    end_time: sunrise + 00:30:00
    end_time_action: "off"
```
In AL, I defined them individually as that was when I had group.bedroom_lights.
![image](https://user-images.githubusercontent.com/49846893/101115877-efd5cf00-3598-11eb-9298-dd4ab79964fa.png)

BTW, any ideas why I have a field sizing issue in the options window?

**Comment by profesaurus** [comment]
@basnijholt & @danielbrunt57, I've got some logs for you.  I have two AL setups.  As I mentioned in my previous message, in my main AL setup I unchecked the ZHA groups from the settings and checked the individual lights.  I haven't seen any issues with this one since doing that.

But I have another AL setup where I left the ZHA light groups in the settings and that one just had an issue with lights turning on after being turned off.  In the settings the light.kids_bathroom_lights is the group I'm referring to:

![image](https://user-images.githubusercontent.com/6362518/101235266-030d9b00-367c-11eb-92eb-1ba5a27200e5.png)
![image](https://user-images.githubusercontent.com/6362518/101235311-9515a380-367c-11eb-859f-653593dc7ff2.png)


I manually turned this group off in HA and they almost immediately came back on:

![image](https://user-images.githubusercontent.com/6362518/101235234-a316f480-367b-11eb-9c91-2577ef1c15cc.png)

These lights have no automations attached to them, so there's nothing I did in HA that could be turning these lights on.

Here are the relevant logs:
> 2020-12-04 21:34:43 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_off('['light.kids_bathroom_lights']', transition=None)' event with context.id='86b79a45c997779df8683d5efa08b3b3'
2020-12-04 21:34:44 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.kids_bathroom_lights' 'state_changed' event: '{'brightness': 5, 'off_brightness': None, 'friendly_name': 'Kids Bathroom Lights', 'supported_features': 41}' with context.id='86b79a45c997779df8683d5efa08b3b3'
2020-12-04 21:34:48 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.kids_bathroom_lights' 'state_changed' event: '{'brightness': 5, 'off_brightness': None, 'friendly_name': 'Kids Bathroom Lights', 'supported_features': 41}' with context.id='86b79a45c997779df8683d5efa08b3b3'
2020-12-04 21:34:48 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Kids Adaptive Lighting: Detected an 'off' → 'on' event for 'light.kids_bathroom_lights' with context.id='86b79a45c997779df8683d5efa08b3b3'
2020-12-04 21:34:48 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Waiting with adjusting 'light.kids_bathroom_lights' for 0.746035
2020-12-04 21:34:49 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.kids_bathroom_lights' 'state_changed' event: '{'brightness': 4, 'off_brightness': None, 'friendly_name': 'Kids Bathroom Lights', 'supported_features': 41}' with context.id='06ecf7490dfee424bb4057cc7d1ae373'

Please let me know if you need any further info.  I'm happy to help.

**Comment by profesaurus** [comment]
I posted too soon.  Looking through my logbook I discovered that my main AL setup had another issue today.  This one is a bit weirder.

I had an automation turn off a ZHA group, light.master_closet_lights, then about 20 seconds later 2 of the 3 lights in the group (light.master_closet_1 & light.master_closet_3) turned on.  These lights are in the same light fixture, so it's not like two can be manually switched on without the others.  Here are the logbook entries:

![image](https://user-images.githubusercontent.com/6362518/101235643-ede63b80-367e-11eb-925a-175e8bb22285.png)
![image](https://user-images.githubusercontent.com/6362518/101235651-05bdbf80-367f-11eb-8c70-aab76c04c7f7.png)
![image](https://user-images.githubusercontent.com/6362518/101235653-08201980-367f-11eb-9a01-974caa28a44f.png)
![image](https://user-images.githubusercontent.com/6362518/101235655-0a827380-367f-11eb-8021-10ad0b1d094c.png)

And the logs:

> 2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_3' 'state_changed' event: '{'brightness': 254, 'off_brightness': None, 'friendly_name': 'Master Closet Light 3', 'supported_features': 41}' with context.id='49aebeb61a3a36006be15fd9e6d8fdc8'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Detected an 'off' → 'on' event for 'light.master_closet_3' with context.id='49aebeb61a3a36006be15fd9e6d8fdc8'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt_e2cf_light_event_80'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Adaptive Lighting: '_adapt_lights(['light.master_closet_3'], 1.0, force=True, context.id=adapt_lgt_e2cf_light_event_80)' called
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.master_closet_3', 'transition': 1.0, 'brightness': 168} with context.id='adapt_lgt_e2cf_light_event_80'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_1' 'state_changed' event: '{'brightness': 254, 'off_brightness': None, 'friendly_name': 'Master Closet Light 1', 'supported_features': 41}' with context.id='a80bb0aabe5af9d5224ea1d10c18c8ed'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.master_closet_3']')' event with context.id='adapt_lgt_e2cf_light_event_80'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Detected an 'off' → 'on' event for 'light.master_closet_1' with context.id='a80bb0aabe5af9d5224ea1d10c18c8ed'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt_e2cf_light_event_81'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Adaptive Lighting: '_adapt_lights(['light.master_closet_1'], 1.0, force=True, context.id=adapt_lgt_e2cf_light_event_81)' called
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.master_closet_1', 'transition': 1.0, 'brightness': 168} with context.id='adapt_lgt_e2cf_light_event_81'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.master_closet_1']')' event with context.id='adapt_lgt_e2cf_light_event_81'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_3' 'state_changed' event: '{'brightness': 168, 'off_brightness': None, 'friendly_name': 'Master Closet Light 3', 'supported_features': 41}' with context.id='adapt_lgt_e2cf_light_event_80'
2020-12-04 17:50:28 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_1' 'state_changed' event: '{'brightness': 168, 'off_brightness': None, 'friendly_name': 'Master Closet Light 1', 'supported_features': 41}' with context.id='adapt_lgt_e2cf_light_event_81'
2020-12-04 17:50:29 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_2' 'state_changed' event: '{'brightness': 254, 'off_brightness': None, 'friendly_name': 'Master Closet Light 2', 'supported_features': 41}' with context.id='093dc58985c4bf33963b210c8021e8a3'
2020-12-04 17:50:29 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Detected an 'off' → 'on' event for 'light.master_closet_2' with context.id='093dc58985c4bf33963b210c8021e8a3'
2020-12-04 17:50:29 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt_e2cf_light_event_82'
2020-12-04 17:50:29 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Adaptive Lighting: '_adapt_lights(['light.master_closet_2'], 1.0, force=True, context.id=adapt_lgt_e2cf_light_event_82)' called
2020-12-04 17:50:29 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.master_closet_2', 'transition': 1.0, 'brightness': 168} with context.id='adapt_lgt_e2cf_light_event_82'
2020-12-04 17:50:29 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.master_closet_2']')' event with context.id='adapt_lgt_e2cf_light_event_82'
2020-12-04 17:50:29 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_1' 'state_changed' event: '{'brightness': 174, 'off_brightness': None, 'friendly_name': 'Master Closet Light 1', 'supported_features': 41}' with context.id='adapt_lgt_e2cf_light_event_81'
2020-12-04 17:50:29 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] State change event of 'light.master_closet_1' is already in 'self.last_state_change' (adapt_lgt_e2cf_light_event_81) adding this state also
2020-12-04 17:50:29 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_2' 'state_changed' event: '{'brightness': 168, 'off_brightness': None, 'friendly_name': 'Master Closet Light 2', 'supported_features': 41}' with context.id='adapt_lgt_e2cf_light_event_82'
2020-12-04 17:50:30 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_1' 'state_changed' event: '{'brightness': 168, 'off_brightness': None, 'friendly_name': 'Master Closet Light 1', 'supported_features': 41}' with context.id='adapt_lgt_e2cf_light_event_81'
2020-12-04 17:50:30 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] State change event of 'light.master_closet_1' is already in 'self.last_state_change' (adapt_lgt_e2cf_light_event_81) adding this state also

> 2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_1' 'state_changed' event: '{'brightness': 167, 'off_brightness': None, 'friendly_name': 'Master Closet Light 1', 'supported_features': 41}' with context.id='7cb77154a2817828037d33c85423661a'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Detected an 'off' → 'on' event for 'light.master_closet_1' with context.id='7cb77154a2817828037d33c85423661a'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt_e2cf_light_event_89'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Adaptive Lighting: '_adapt_lights(['light.master_closet_1'], 1.0, force=True, context.id=adapt_lgt_e2cf_light_event_89)' called
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.master_closet_1', 'transition': 1.0, 'brightness': 166} with context.id='adapt_lgt_e2cf_light_event_89'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.master_closet_1']')' event with context.id='adapt_lgt_e2cf_light_event_89'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_1' 'state_changed' event: '{'brightness': 166, 'off_brightness': None, 'friendly_name': 'Master Closet Light 1', 'supported_features': 41}' with context.id='adapt_lgt_e2cf_light_event_89'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_3' 'state_changed' event: '{'brightness': 167, 'off_brightness': None, 'friendly_name': 'Master Closet Light 3', 'supported_features': 41}' with context.id='00ef699d40e73439198ae6e629f7dc8d'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Detected an 'off' → 'on' event for 'light.master_closet_3' with context.id='00ef699d40e73439198ae6e629f7dc8d'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt_e2cf_light_event_90'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Adaptive Lighting: '_adapt_lights(['light.master_closet_3'], 1.0, force=True, context.id=adapt_lgt_e2cf_light_event_90)' called
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.master_closet_3', 'transition': 1.0, 'brightness': 166} with context.id='adapt_lgt_e2cf_light_event_90'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.master_closet_3']')' event with context.id='adapt_lgt_e2cf_light_event_90'
2020-12-04 17:54:18 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.master_closet_3' 'state_changed' event: '{'brightness': 166, 'off_brightness': None, 'friendly_name': 'Master Closet Light 3', 'supported_features': 41}' with context.id='adapt_lgt_e2cf_light_event_90'

**Comment by Chreece** [comment]
I can also confirm that with my Osram Zigbee lamp

**Comment by willtho89** [comment]
I did some testing and it looks like the problem is as suggested by @basnijholt: The states of ZHA devices is a bit unreliable. If a device is seen as on in HA, adaptive light tries to adjust its brightness/temperature according to your settings.  Unfortunately lights turn on on brightness change. This seems to be an implementation in HA itself.

@basnijholt i think it would be reasonable to add a flag `ignore off -> on state change for 1 intervall (or x sec)`. This might mitigate the error here

**Comment by miezie** [comment]
I can confirm the same behavior with Philips and Ikea lights on a Conbee2 stick with Deconz. More then with the original circadian lighting component. Think it relates to the combination of Deconz groups, single light bulbs and HA light groups. Will make that more sophisticated on my side but would like to share my initial thought...

**Comment by drinfernoo** [comment]
Has anyone been able to figure out what is happening here? I have two Sengled bulbs which seem to be doing something similar to this... where they *sometimes* come on randomly, but only at night. It started with one of my bulbs, and has been happening off and on (sometimes doesn't happen at all, sometimes happens many times in a night) for a few months. It just started last night with the second bulb.

Both bulbs are bound directly to one Lutron Aurora each, and none of my Zigbee devices are part of any group (HA or Zigbee). I use ZHA with an HZUSBZ-1 stick.

**Comment by miezie** [comment]
> I can confirm the same behavior with Philips and Ikea lights on a Conbee2 stick with Deconz. More then with the original circadian lighting component. Think it relates to the combination of Deconz groups, single light bulbs and HA light groups. Will make that more sophisticated on my side but would like to share my initial thought...

I did my checkings and refactored it entirely... Using groups or not using them doesn't differ. I do have one positive thing here... it happens quite a lot and most of the time when shutting down several entities. I will raise loglevels to give some valuable input later on.

**Comment by broyuken** [comment]
I seem to have been able to fix it. I think the problem comes in if a light is transitioning while it is shut off, the transition continues which turns the light back on. I fixed this by setting the transition to 1 second. Haven’t had the issue since.

**Comment by lennartgrunau** [comment]
> I seem to have been able to fix it. I think the problem comes in if a light is transitioning while it is shut off, the transition continues which turns the light back on. I fixed this by setting the transition to 1 second. Haven’t had the issue since.

Trying this right now. 

Additionally, I created an automation that fires a `light.turn_off` ~10 seconds after the first turn off.
Elegant? Nah, but it definitly turns the lights off, when I leave home.

**Comment by willtho89** [comment]
> I seem to have been able to fix it. I think the problem comes in if a light is transitioning while it is shut off, the transition continues which turns the light back on. I fixed this by setting the transition to 1 second. Haven’t had the issue since.  

Doesn't work for me unfortunately. Problem seems to be the slow refresh of zigbee lights (thinking light is still on)

**Comment by miezie** [comment]
Not using separate_turn_on_commands fixed it for me with Hue and Ikea lights.

**Comment by jd1900** [comment]
> > I seem to have been able to fix it. I think the problem comes in if a light is transitioning while it is shut off, the transition continues which turns the light back on. I fixed this by setting the transition to 1 second. Haven’t had the issue since.
> 
> Doesn't work for me unfortunately. Problem seems to be the slow refresh of zigbee lights (thinking light is still on)

Yes, I think the problem is related to a delay from the zigbee network and HA, specially when controlled outside HA (for instance binding devices).

I think this is very related to the this issue:
https://github.com/basnijholt/adaptive-lighting/issues/96

IMHO, disabling brightness control and updating temperature not "turning on" the device (just changing temperature) could fix it.

**Comment by danielbrunt57** [comment]
I developed this issue about 1-2 weeks ago with my office & bedroom lights.  Would appear to have been a fight between adaptive lighting and magic areas.  I disabled AL for office and the lights stopped coming back on after magic area turned them off.  A few upgrades later, re-enabled AL for office and they seem to be okay again now...

**Comment by danielbrunt57** [comment]
I spoke to soon.  It JUST happened again...

**Comment by dieneuser** [comment]
I have the exact same problem / behavior. I use ZHA first with an Silicon Labs EmberZNet based coordinator (belows) and now a Conbee 2 (deconz) coordinator. Same behaviour like describes with random lamps of all kinds in my automation (Ikea, Philips, innr, Paulmann, Osram). Regardless of whether these are grouped in ZHA or not. Only turning the AL switch off solves the problem permanently.
That is really very annoying - otherwise the integration is great.

**Comment by fpschrisiom** [comment]
Same issue for me, ZHA with conbee2. I have aurora dimmer, dumb lamps.
Only happens when the adaption is running and light turned off before it completes.
Usually turns on full brightness in the day as I tend to override turning the light off although I have managed to turn off the lights in the evening before fully dim and they've turned back on.

**Comment by arnefm** [comment]
Same here. ZHA with Conbee 2. Four Hue bulbs set up in a ZHA zigbee group. After installing and setting up Adaptive Lighting I'm unable to turn them off permanently in the evening. They're controlled by an automation calling light.turn_off after a 5 minutes of no motion. I can also turn them off manually with my bedside Hue dimmer. They consistently turn back on after a few seconds. 

WAF took a huge hit last night after my wife triggered the motion sensor and could not get the lights to turn off again at all. They stayed on all night.

**Comment by shaner-git** [comment]
I also had this issue since implementing adaptive lighting. I think I finally found the culprit. 

What I did:

- Going into HA > Configuration > Integrations > Adaptive Lighting, then selecting the group that is turning back on
- Select Options
- Find "Transition" then drop that value down to something minimal, like 1. 

I haven't played around with other values, this may be able to be higher than 1, but I've confirmed a transition set to 1 will prevent the light that just turned off from turning back on immediately after.

**Comment by arnefm** [comment]
> I also had this issue since implementing adaptive lighting. I think I finally found the culprit.
> 
> What I did:
> 
>     * Going into HA > Configuration > Integrations > Adaptive Lighting, then selecting the group that is turning back on
> 
>     * Select Options
> 
>     * Find "Transition" then drop that value down to something minimal, like 1.
> 
> 
> I haven't played around with other values, this may be able to be higher than 1, but I've confirmed a transition set to 1 will prevent the light that just turned off from turning back on immediately after.

Okay, so that supports the theory that this bug occurs when the lights are turned off during an ongoing transition - Lowering the transition time reduces the likelihood of this happening.

I'll definitely try lowering the transition time and find a good compromise between transition time and interval. As long as the transitions aren't too noticeable that should work fine.

Of course, there's will still be a chance that this bug occurs but it should be rare.

**Comment by danielbrunt57** [comment]
So that's the issue!  I just finished migrating away from Magic Areas and Adaptive Lighting to a cool Circadian subroutine and Node-RED. I finally delved into NR a week ago and after many MANY hours I have my motion routines setup and circadian implanted where I have dimmable, color controlled bulbs.

**Comment by danielbrunt57** [comment]
So I've spent what seems like an eternity inside Node_RED implementing my own version of Magic Areas and Adaptive Lighting using code from [Circadian/Adaptive Lighting NodeRed](https://community.home-assistant.io/t/circadian-adaptive-lighting-nodered/269741).  I've experienced the light(s)s turning back on if turned off within the transition period within my Node-RED implementation.  Bedroom Lamp 1 (Sengled Zigbee) was doing it.  It was part of a zigbee group bedroom_lights with bedroom_lamp_1 and bedroom_lamp2.  I removed the group and experimented with the individual lights and only lamp_1 has the problem, not lamp_2, **BUT** only when turning it off from the HA UI!  If I turn it off from a button injection in Node-RED, no problem.  So, the issue would seem to lie somewhere (deep) within HA and **_only_** when HA is controlling the light AND seems to be device dependent since lamp 2 is not affected.  I have 5 of these bulbs and all have the same firmware so its not that.  I've not played yet with the 3 bulbs in the office or the dimmer switch for the bathroom lights where I also saw random turn-ons...

![image](https://user-images.githubusercontent.com/49846893/113107666-0b548d80-91b9-11eb-9361-06b00869c6d7.png)

![image](https://user-images.githubusercontent.com/49846893/113107760-232c1180-91b9-11eb-8712-b1dff375040e.png)

**Comment by danielbrunt57** [comment]
It's also an issue from here...

![image](https://user-images.githubusercontent.com/49846893/113109583-f842bd00-91ba-11eb-9007-47bb674cff2f.png)

**Comment by danielbrunt57** [comment]
I also tested with two different browser views into HA...one where I called service light.turn_on with a transition of 20s and one with light.turn_off and was able to reproduce the problem WITH MY NR CIRCADIAN DISABLED!

![image](https://user-images.githubusercontent.com/49846893/113113009-b156c680-91be-11eb-859a-b85cfd06ddf7.png)

![image](https://user-images.githubusercontent.com/49846893/113112335-eca4c580-91bd-11eb-99f5-d1628f822141.png)

![image](https://user-images.githubusercontent.com/49846893/113112437-06460d00-91be-11eb-953e-f9d87a301fa5.png)

Lamp 1 came back on...
![image](https://user-images.githubusercontent.com/49846893/113112560-27a6f900-91be-11eb-8ac5-31458faea70d.png)

**Comment by basnijholt** [comment]
Wow, @danielbrunt57! Thanks a lot for your excellent analysis. I already had a suspicion that this wasn't related to Adaptive-lighting itself because nothing in the logs indicates that this should happen.

What do you think would be your suggestion to tackle this problem?

**Comment by danielbrunt57** [comment]
I think we wait for HA to review issue #48534.  
It is very late here but it kind of looks like an issue with core to me.  Since posting I did experience a lamp_2 turn on using a transition of 0 and also no transition in my circadian routine so it's possible it might be something in ZHA since only lamp_1 is turning back on and not lamp_2.  Strange though that it's only when the service is called straight from HA.  If I trigger `off` from NR it does not come back on...

**Comment by danielbrunt57** [comment]
The issue **seems** to occur when the brightness_pct is different from the previous one.
using`transition: 20`

9:26:00 turn_on transition: 20 brightness_pct: 80
9:26:00 turn_off

9:27:45 turn_on transition: 20 brightness_pct: 90
9:27:50 turn_on
9:28:05 **turned on by itself**

9:29:00 turn_on transition: 20 brightness_pct: 90
9:29:05 turn_off

9:30:00 turn_on transition: 20 brightness_pct: 70
9:30:05 turn_off
9:30:20 **turned on by itself**

**Comment by filikun** [comment]
So I had this happen to me for the first time today. I have been using this integration for a long while but decided to start fresh and setup a new HA instance. Now I have not read all the posts in here but my first reaction was to check my groups. 

So my scenario is from the living room. I have AL on the light group for all my living room lights (not on the lights themselves) and I do have a group that just contains the lights I want to turn off when the tv is on. So when light.livingroom is still technically on even when light.tv is off so I think it applied the transition to the turned off tv lights as AL was set to use the light.livingroom and that’s why the tv lights turns on even when they should not. 

I tested it out and put the light.livingroom to my tv automation instead of just turning off them tv lights and the problem was gone. Not sure if it’s related but that’s my finding.

**Comment by sevorl** [comment]
I use deconz with ikea bulbs and ikea 5 way buttons. 
a few seconds after I turn off a group of lights using the ikea button, and one or two random lights will turn back on. 
Is this there any solution for this yet?

**Comment by danielbrunt57** [comment]
@sevorl The problem is not Adaptive Lighting!  It ONLY controls lights that show as ON in Home Assistant.

Home Assistant is still showing the light as ON when it is physically off.
Disable AL, sit in front of HA and observe the state of the lights as you turn them off with your buttons.
I have, and I've seen it with my two groups of Sengled Zigbee bulbs.
Another anomaly I've seen is using a transition with a group can also cause HA to get out of sync.  It depends on a lot of factors one of which may be that while transitioning to off, the light(s) report an intermediate brightness state causing HA to toggle its state back to on, then the light reaches off and that message gets lost somewhere...

**Comment by broyuken** [comment]
I've had lights randomly go on in the middle of the night, with nothing to trigger them. Not saying it's AL, but I never had this happen until I started using AL.

**Comment by sevorl** [comment]
When I turn lights off with the switch, they will show as off in ha immediately. Also, this did not happen before I started using AL.  I will test further though to see if I can reproduce this when AL is off, or if it would disappear without transitioning time.

**Comment by RubenKelevra** [comment]
@broyuken I feel like this might be a different problem. Can you create a dedicated ticket for this - this way it won't get lost. :)

**Comment by RubenKelevra** [comment]
I think the original issue might be just a race condition: A light might slowly turn off, so it will take some seconds before it's at 0 brightness. In this time it is reported as "on" and will only report "off" when it truly turned off.

If adaptive lighting is pushing an update command while the light is in the transition it will abort the transition to off, and execute the new command.

I think there are two workarounds:

- You can create an button which executes a script, which turns the lights off, when they are on, waits some seconds and does this again. This way it might take longer to turn off the light, but it will be retried if the first time failed.
- I'm using Zigbee2mqtt. This software has a setting that commands sent to devices are handled "optimistically". This means if you send a "turn off" command to the light, Zigbee2mqtt will send a "light's off" back, before the device has reacted. This way you don't have to wait for the eventual zigbee latency if there's a lot of traffic or you turn off many lights at the same time. Since adaptive lighting gets the change faster there's a much lower chance that a race condition happens.

**Comment by RubenKelevra** [comment]
I'm currently in the process of doing version 2. This will be added/fixed.

Keep this ticket open until you can verify that it's properly working in the upcoming version 2.

**Comment by basnijholt** [comment]
This problem started to occur in my own home after upgrading the firmware of my Shelly Dimmer 2 switches.

I turned off the lights using a Siri (HomeKit) voice command:
```sh
2022-11-18 19:41:18.717 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_off('['light.ceiling_kitchen']', transition=1)' event with context.id='01GJ6ZKWFR1JKVNXTSN96H7QFF'
```
11 seconds later, the light reported back to Home Assistant that its state was on, which triggers a *'off' → 'on' event* in Home Assistant:
```sh
2022-11-18 19:41:29.092 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.ceiling_kitchen' 'state_changed' event: '{'supported_color_modes': [<ColorMode.BRIGHTNESS: 'brightness'>], 'color_mode': <ColorMode.BRIGHTNESS: 'brightness'>, 'brightness': 204, 'friendly_name': 'Ceiling kitchen', 'supported_features': 32}' with context.id='01GJ6ZM6M26G80GVMYQFYQK716'
2022-11-18 19:41:29.093 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Detected an 'off' → 'on' event for 'light.ceiling_kitchen' with context.id='01GJ6ZM6M26G80GVMYQFYQK716'
```
Then immediately, Adaptive Lighting interprets this as a light that turned on and starts adapting the brightness and color ever so slightly:
```sh
2022-11-18 19:41:29.094 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'M8)K':light_event:b'!xR''
2022-11-18 19:41:29.096 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: default: '_adapt_lights(['light.ceiling_kitchen'], 1.0, force=True, context.id=adapt_lgt:b'M8)K':light_event:b'!xR')' called
2022-11-18 19:41:29.100 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.ceiling_kitchen', 'transition': 1.0, 'brightness': 203} with context.id='adapt_lgt:b'M8)K':light_event:b'!xR''
2022-11-18 19:41:29.102 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.ceiling_kitchen']')' event with context.id='adapt_lgt:b'M8)K':light_event:b'!xR''
```

This means that the light still tells Home Assistant that is was on, even though it is off! I believe an incorrect reporting of the lights is the root cause of this issue. At least in my case.


The full log around that time:

```bash
2022-11-18 19:41:18.717 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_off('['light.ceiling_kitchen']', transition=1)' event with context.id='01GJ6ZKWFR1JKVNXTSN96H7QFF'
2022-11-18 19:41:18.719 WARNING (MainThread) [homeassistant.helpers.service] Unable to find referenced entities light.kitchen_counter or it is/they are currently not available
2022-11-18 19:41:19.353 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.cabinet_led' 'state_changed' event: '{'min_color_temp_kelvin': 1538, 'max_color_temp_kelvin': 7142, 'min_mireds': 140, 'max_mireds': 650, 'effect_list': ['colorloop'], 'supported_color_modes': [<ColorMode.COLOR_TEMP: 'color_temp'>, <ColorMode.HS: 'hs'>, <ColorMode.XY: 'xy'>], 'color_mode': <ColorMode.XY: 'xy'>, 'brightness': 203, 'hs_color': (28.673, 88.627), 'rgb_color': (255, 137, 29), 'xy_color': (0.597, 0.3793), 'is_deconz_group': False, 'friendly_name': 'Cabinet LED', 'supported_features': 44}' with context.id='01GJ6ZKX3NHD0RHQHSADG7Z0SC'
2022-11-18 19:41:29.092 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.ceiling_kitchen' 'state_changed' event: '{'supported_color_modes': [<ColorMode.BRIGHTNESS: 'brightness'>], 'color_mode': <ColorMode.BRIGHTNESS: 'brightness'>, 'brightness': 204, 'friendly_name': 'Ceiling kitchen', 'supported_features': 32}' with context.id='01GJ6ZM6M26G80GVMYQFYQK716'
2022-11-18 19:41:29.093 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Detected an 'off' → 'on' event for 'light.ceiling_kitchen' with context.id='01GJ6ZM6M26G80GVMYQFYQK716'
2022-11-18 19:41:29.094 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'M8)K':light_event:b'!xR''
2022-11-18 19:41:29.096 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: default: '_adapt_lights(['light.ceiling_kitchen'], 1.0, force=True, context.id=adapt_lgt:b'M8)K':light_event:b'!xR')' called
2022-11-18 19:41:29.100 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.ceiling_kitchen', 'transition': 1.0, 'brightness': 203} with context.id='adapt_lgt:b'M8)K':light_event:b'!xR''
2022-11-18 19:41:29.102 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.ceiling_kitchen']')' event with context.id='adapt_lgt:b'M8)K':light_event:b'!xR''
2022-11-18 19:41:33.606 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'M8)K':interval:b'#1s''
2022-11-18 19:41:33.608 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: default: '_adapt_lights(['light.ceiling_kitchen', 'light.ceiling_bedroom', 'light.bed_reading', 'light.hall_lights', 'light.bathroom_ceiling', 'light.lampan', 'light.philips_go', 'light.bamboo', 'light.toilet'], 45.0, force=False, context.id=adapt_lgt:b'M8)K':interval:b'#1s')' called
2022-11-18 19:41:33.609 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.ceiling_kitchen', 'transition': 45.0, 'brightness': 203} with context.id='adapt_lgt:b'M8)K':interval:b'#1s''
2022-11-18 19:41:33.610 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.ceiling_kitchen']')' event with context.id='adapt_lgt:b'M8)K':interval:b'#1s''
2022-11-18 19:41:33.997 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default bed: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'z_O?':interval:b'J`?''
2022-11-18 19:41:33.999 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: default bed: '_adapt_lights(['light.bed_led'], 45.0, force=False, context.id=adapt_lgt:b'z_O?':interval:b'J`?')' called
```

**Comment by basnijholt** [comment]
Another incident, same story:

Home Assistant turned off my light here:
```sh
2022-11-18 21:44:04.816 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_off('['light.ceiling_kitchen']', transition=1)' event with context.id='01GJ76MNYBXF5EDBJF8F83SN04'
```
again 11 seconds later, the light reported back to Home Assistant that its state was on, which triggers a *'off' → 'on'* event in Home Assistant:
```
2022-11-18 21:44:15.200 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.ceiling_kitchen' 'state_changed' event: '{'supported_color_modes': [<ColorMode.BRIGHTNESS: 'brightness'>], 'color_mode': <ColorMode.BRIGHTNESS: 'brightness'>, 'brightness': 184, 'friendly_name': 'Ceiling kitchen', 'supported_features': 32}' with context.id='01GJ76N02YJKDGG5GBT8X10S4K'
2022-11-18 21:44:15.201 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Detected an 'off' → 'on' event for 'light.ceiling_kitchen' with context.id='01GJ76N02YJKDGG5GBT8X10S4K'
```
Then immediately, Adaptive Lighting interprets this as a light that turned on and starts adapting the brightness and color ever so slightly:
```sh
2022-11-18 21:44:15.202 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'M8)K':light_event:b'#1#''
2022-11-18 21:44:15.203 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: default: '_adapt_lights(['light.ceiling_kitchen'], 1.0, force=True, context.id=adapt_lgt:b'M8)K':light_event:b'#1#')' called
2022-11-18 21:44:15.208 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.ceiling_kitchen', 'transition': 1.0, 'brightness': 185} with context.id='adapt_lgt:b'M8)K':light_event:b'#1#''
2022-11-18 21:44:15.212 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.ceiling_kitchen']')' event with context.id='adapt_lgt:b'M8)K':light_event:b'#1#''
```


The full logs:
```sh

2022-11-18 21:44:04.816 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_off('['light.ceiling_kitchen']', transition=1)' event with context.id='01GJ76MNYBXF5EDBJF8F83SN04'
2022-11-18 21:44:15.200 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected a 'light.ceiling_kitchen' 'state_changed' event: '{'supported_color_modes': [<ColorMode.BRIGHTNESS: 'brightness'>], 'color_mode': <ColorMode.BRIGHTNESS: 'brightness'>, 'brightness': 184, 'friendly_name': 'Ceiling kitchen', 'supported_features': 32}' with context.id='01GJ76N02YJKDGG5GBT8X10S4K'
2022-11-18 21:44:15.201 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Detected an 'off' → 'on' event for 'light.ceiling_kitchen' with context.id='01GJ76N02YJKDGG5GBT8X10S4K'
2022-11-18 21:44:15.202 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'M8)K':light_event:b'#1#''
2022-11-18 21:44:15.203 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: default: '_adapt_lights(['light.ceiling_kitchen'], 1.0, force=True, context.id=adapt_lgt:b'M8)K':light_event:b'#1#')' called
2022-11-18 21:44:15.208 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.ceiling_kitchen', 'transition': 1.0, 'brightness': 185} with context.id='adapt_lgt:b'M8)K':light_event:b'#1#''
2022-11-18 21:44:15.212 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.ceiling_kitchen']')' event with context.id='adapt_lgt:b'M8)K':light_event:b'#1#''
2022-11-18 21:44:33.847 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'M8)K':interval:b'#T5''
2022-11-18 21:44:33.849 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: default: '_adapt_lights(['light.ceiling_kitchen', 'light.ceiling_bedroom', 'light.bed_reading', 'light.hall_lights', 'light.bathroom_ceiling', 'light.lampan', 'light.philips_go', 'light.bamboo', 'light.toilet'], 45.0, force=False, context.id=adapt_lgt:b'M8)K':interval:b'#T5')' called
2022-11-18 21:44:33.851 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default: Scheduling 'light.turn_on' with the following 'service_data': {'entity_id': 'light.ceiling_kitchen', 'transition': 45.0, 'brightness': 185} with context.id='adapt_lgt:b'M8)K':interval:b'#T5''
2022-11-18 21:44:33.852 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Detected an 'light.turn_on('['light.ceiling_kitchen']')' event with context.id='adapt_lgt:b'M8)K':interval:b'#T5''
2022-11-18 21:44:34.222 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] default bed: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'z_O?':interval:b'HWd''
2022-11-18 21:44:34.224 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: default bed: '_adapt_lights(['light.bed_led'], 45.0, force=False, context.id=adapt_lgt:b'z_O?':interval:b'HWd')' called
2022-11-18 21:44:38.991 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] test: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'%t+n':interval:b'ga-''
2022-11-18 21:44:43.574 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] living room: '_update_attrs_and_maybe_adapt_lights' called with context.id='adapt_lgt:b'-DM|':interval:b'4Hp''
2022-11-18 21:44:43.577 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Adaptive Lighting: living room: '_adapt_lights(['light.tv_led', 'light.cabinet_led', 'light.corner_living_room', 'light.sphere'], 45.0, force=False, context.id=adapt_lgt:b'-DM|':interval:b'4Hp')' called
```

**Comment by basnijholt** [comment]
One potential solution is to **optionally** _always ignore state changes of the light_, and only listen to `light.turn_on` and `light.turn_off` events.

This means that Adaptive Lighting will _not_ kick in when you turn on the light _outside of Home Assistant_.

**Comment by basnijholt** [comment]
OK lol...

What I just proposed, I already implemented 😂 

Can folks try to set `detect_non_ha_changes` to False and see whether this still happens?

**Comment by GollyJer** [comment]
> One potential solution is to **optionally** _always ignore state changes of the light_, and only listen to `light.turn_on` and `light.turn_off` events.
> 
> This means that Adaptive Lighting will _not_ kick in when you turn on the light _outside of Home Assistant_.

Hey @basnijholt. 
Does "outside of home assistant" mean with voice assistants and timers? So only turning the lights on with the Home Assistant app will trigger Adaptive Lighting?

**Comment by danielbrunt57** [comment]
> Does "outside of home assistant" mean with voice assistants and timers? So only turning the lights on with the Home Assistant app will trigger Adaptive Lighting? Just want to be sure I have the right expectations for testing.

I would say no *if* the voice assistant is using Home Assistant to trigger the change.  If the voice assistant is using a skill linked to a third party web interface like Ewelink, Tuya, Kasa, etc. then that change would be considered outside of HA.
As for timers(?), that sounds like outside of HA...

**Comment by danielbrunt57** [comment]
One point I would like to make is that some dimmable devices will report 'on' if a timed transition from 'on' to 'off' is used.  They (not all) seem to report 'on/40%', 'on/30%', 'on/20%' etc. and I think this could be a factor also...

**Comment by danielbrunt57** [comment]
FEI, since I abandoned (but not forgotten!) AL my lights have always worked properly.  It was a lot more work but, in the end, more reliable.  I hope this issue gets resolved!!!

**Comment by daFreeMan** [comment]
@basnijholt if by "set `detect_non_ha_changes` to False and see whether this still happens?" you mean this:
![image](https://user-images.githubusercontent.com/11889733/203152514-ca0b9ab7-6b12-4ca1-8b1c-fe6b9993d543.png)
then I've had it set that way almost the whole time I've used AL and still had this issue. If there's another setting, please let me know and I'll try that.

However, as noted on the now closed #343, I am _not_ having the issue on release 1.2.1.

@danielbrunt57 , curious if you've found an alternative solution to AL, or if you're just adjusting manually. Do note, I've not had the issue since updating to 1.2.1

**Comment by pimeys** [comment]
Also having the same issue. It happens when I set AL to work with ZigBee groups. The error is not happening if I add AL to the lights separately without using a group.

Of course if I do not use groups, sometimes all messages do not get to all the lights and I might have one light in a wrong color for a while compared to the rest of the group.

**Comment by danielbrunt57** [comment]
I found the same thing with ZigBee groups. I think the issue stems from the fact that processing messages is serial and sequential and boils down to timing. It's virtually impossible IMO to process in parallel and have all statuses show off immediately after they've been turned off. There are numerous points at which processing delays happen. In my own circadian routines in NR I had to be very careful and had to add delays and double checks. Programmers have battled parallel processing for decades.  It's a tough one to solve.

**Comment by pimeys** [comment]
@danielbrunt57 do you think there's some nice `sleep` that could be useful here? The best of course would be a separate message that just changes the color and dims, but doesn't turn on the light. Sadly this does not exist.

**Comment by pimeys** [comment]
I downgraded to 1.0.20, which seems to work better but not perfectly still. It really looks like we have a race condition in the system. When turning off the lights, it takes a while sometimes until they report being off. So if at that moment adaptive lightning decides it's time to change the color, it might turn on some of the lights. With the latest version I also had an issue of lights turning on in the middle of the night, but this old version only has a bug sometimes when you turn off the lights from the switch, and they might turn on a few seconds afterwards.

I tried to reduce the issue by increasing the interval from 90 seconds to 300 seconds. Hopefully this way I hit the race condition much more rarely.

**Comment by danielbrunt57** [comment]
Unrelated to AL but along the lines of "race conditions"...
I just resolved an issue today that I have had for weeks now with my bathroom light coming back on but only when turned off via my good_night.yaml script. The script first turns on sleep_mode followed by all lights off except bedroom.  2 seconds later the bathroom light was coming back on.  Node-RED/HA timing delay was the problem.  My bathroom light motion control is done in Node-RED which also uses sleep_mode to dim the bathroom light.  The problem was that `sleep_mode on` was triggering NR to dim the light.  Once triggered, it did first check if the light was on, which it still saw as on at that point, but by the time it got to the service light.turn on, HA had switched the light off and so the light came back on. I had to insert a 1s delay between sleep mode and is_light_on for it to not turn the light back on.
![image](https://user-images.githubusercontent.com/49846893/204692008-26a1e4aa-e04e-48b8-a54c-709affe05cea.png)

I think it serves to illustrate that minor delays in transfer or recognition of states between different components can lead to problems, especially when trying to parallel process things.  "The brain tells the left hand what the right hand is doing but by the time that occurs the left hand is doing something else."

It took me a while to trace it as the HA logs only showed `Turn on triggered by state of Bathroom Light turned off - Supervisor` WTH?  What does that even mean?
![image](https://user-images.githubusercontent.com/49846893/204693710-ea220d75-a47e-456c-8c3c-d34a43c6b1d1.png)

**Comment by th3w1zard1** [comment]
This has been a long-standing issue that I've been trying to solve before I even found this light adapting project two years ago.

The main problem with the current implementation is it calls async_update_entity every 'interval' as stated in the docs. While this works for most lights this function call can actually cause a light to update its status as homeassistant will reach out to the zigbee network for the most recent data.

This causes adaptive_lighting to detect all of the garbage updates that entityupdate produces. You'd think this wouldn't be a problem because at least it's noticing a change right? Well adaptive_lighting checks for **significant** changes in brightness/color temp, creating a very realistic scenario that a lightchange is detected incorrectly.

^ Please note I could be wrong on the specific root cause of the myriad of scenarios that could cause the problem in that last paragraph but regardless the implementation could be better.

The way I found to implement a better system is to watch for state_changes in home assistant's events, with a simple automation. It's quite a inefficient method as that automation gets called for **_every_** state change not just lights. But it works well for my purposes.

Back on the topic of a potential fix here, my proposal is to add yet another interval for lightchanges:
if adaptive_lighting detects a change, small or big, it should start a timer for _interval_ seconds and wait for the change to stop.
When the change starts, then do the calculations already implemented to detect if it was a significant change: if so set to manual
If the light keeps changing after _interval_ seconds the program should set that light as manual.
Also don't forget to ignore any adapt calls for that light during this interval.

**Comment by danielbrunt57** [comment]
> The way I found to implement a better system is to watch for state_changes in home assistant's events, with a simple automation. It's quite an inefficient method as that automation gets called for **_every_** state change not just lights. But it works well for my purposes.

I'm interested in your automation....

**Comment by th3w1zard1** [comment]
> > The way I found to implement a better system is to watch for state_changes in home assistant's events, with a simple automation. It's quite an inefficient method as that automation gets called for **_every_** state change not just lights. But it works well for my purposes.
> 
> I'm interested in your automation....

[See the readme.md in my fork](https://github.com/th3w1zard1/adaptive_lighting) I have all the automation listed with instructions on how to install them.

**Comment by danielbrunt57** [comment]
> [See the readme.md in my fork](https://github.com/th3w1zard1/adaptive-lighting) I have all the automation listed with instructions on how to install them.
Well, it's not this one...
```
- alias: "Adaptive lighting: toggle 'sleep mode'"
  trigger:
    - platform: state
      entity_id: input_boolean.sleep_mode
'''
and it does not look like the one before...
```
- alias: "Adaptive lighting: reset manual_control after 1 hour"
  mode: parallel
  trigger:
    platform: event
    event_type: adaptive_lighting.manual_control
```
???

**Comment by th3w1zard1** [comment]
@danielbrunt57 My apologies, just refresh, I had to push the updated readme.md off a private branch. It's at the bottom right above Graphs.

**Comment by th3w1zard1** [comment]
Back on topic, I'd like to also point out that the current implementation via 'take_control' and 'detect_non_ha_changes' is very strict. In my opinion, a light being adapted and turned on in the middle of the night is a significantly larger problem than a light being incorrectly set to manual mode. Most users are using the 'reset after 1 hour' automation to reset manual mode anyway.

In addition to my 'interval' suggestion, perhaps a variable could be set similar to manual_mode called 'temp_manual_mode' which will cause the program to update the light's values, wait _interval_ seconds until it's **certain** it can adapt it.

TL;DR there should be more safeguards in place to prevent a light from turning on on its own, a light incorrectly set as 'manual' is a much much smaller issue than a light waking up at 2 am because your smart home is rebelling against the user lol.

**Comment by th3w1zard1** [comment]
Following up with some newfound additional research on this issue.

I realize now that most users who have this problem probably have adaptive-lighting as their **only** automation involving lights. At the very least, adaptive-lighting modify's lights more than any other automation inside of HASSIO.
**Because of this, I believe it's entirely possible adaptive-lighting isn't causing issue #10 at all, and this issue is always caused outside of adaptive-lighting or even user error.**
Please allow me just a moment to explain this happening in a scenario where adaptive-lighting isn't involved:

Setup: user has Zigbee lightbulbs in their bedroom connected to a LAN server running _HASSIO_ and a zigbee mesh running by the addon _'zigbee2mqtt'_. No automations are running.
Lights are connected in the following manner:
`switch >> light >> zigbee2mqtt >> home assistant`. The switch is decidedly not connected to HASSIO or zigbee2mqtt
```
>user has his lights on in his bedroom for most of the day.
>light disconnects momentarily due to a connection hiccup
>user turns bedroom lights off to go to sleep
```

Here's two problematic possibilities:
```
>zigbee2mqtt pings the bulb in the middle of the night since it hasn't heard from it in hours.
>bulb replies
>zigbee2mqtt sends that info to hassio
**>hassio tells bulbs to resume last known state ON.**
OR
>bulb reports its status to hassio in the middle of the night.
**_>zigbee2mqtt replies with its knowledge of the bulb's last known state ON_**
```

These are common scenarios Zigbee light users have to deal with on initial setup. Usually, all the user needs to do is set 'power_on_state' to 'off' in the bulb itself, or fix their MQTT retention settings.
Another common user error that I can't fit into this post, is incorrectly setting up switches to retain messages. Doing this will cause the bulb to execute the switch's command when it inevitably reports to zigbee2mqtt, even if that command was issued hours ago. [See this video for more information on MQTT retention settings](https://www.youtube.com/watch?v=31IyfM1gygo)
Now let's assume the user fixed that during initial installation, then installed adaptive-lighting.

If you're still reading: I thank you for sticking around this far. Let me now explain how this relates to adaptive-lighting and this issue thread #10.

With adaptive-lighting, the dev in his infinite wisdom implemented a very nice and surprisingly simple 'detect_non_ha_changes' feature. It definitely runs correctly in all the tests on my end. We're all very grateful.
Yet randomly, I'll still have bulbs turn themselves on in the middle of the night despite 'detect_non_ha_changes: true' set.

And finally my main point. I can think of two possible causes to issue #10:

### 1. Update_entity causes zigbee2mqtt to ping the bulb. With 'power_on_state: last_known' setup in its device settings AND improper retain options from the switch available on the MQTT broker, zigbee2mqtt would tell the bulb to turn on like it believes it's supposed to be.

### 2. update_entity doesn't actually grab a state report from the bulb, just the data from zigbee2mqtt available in the broker, causing adaptive-lighting to use unreliable information to do its check for a significant change.

The snippet below from switch.py is only called when detect_non_ha_changes is set to true.

```
    async def significant_change(
...
    ) -> bool:
        """Has the light made a significant change since last update.

        This method will detect changes that were made to the light without
        calling 'light.turn_on', so outside of Home Assistant. If a change is
        detected, we mark the light as 'manually controlled' until the light
        or switch is turned 'off' and 'on' again.
        """
        if light not in self.last_state_change:
            return False
        old_states: list[State] = self.last_state_change[light]
        **_await self.hass.helpers.entity_component.async_update_entity(light)_**
        new_state = self.hass.states.get(light)
        ... some compare checks are below ...
```
Scenario 1 would already have its light on right after the update_entity call.

In scenario 2, old_state could have brightness at 255 and new_state according to zigbee2mqtt could be 255. The bulb hasn't even been pinged despite the update_entity call!
What's worse is this might not be apparent in the first day or even week of using adaptive-lighting, just on the scenarios where zigbee2mqtt occasionally has slightly outdated information available.

The **fix** in scenario 2 would be to **set bulb reporting to 'on'**, turn off **retain** in its **device settings**, and ensure zigbee2mqtt is actually pinging the bulb directly for its current state everytime.

Yes I'm talking about zigbee2mqtt which I'm using and many of you might not be, but I think this can happen with any system sending messages through MQTT. Tasmota comes to mind as using MQTT as well. [Again, please see the video](https://www.youtube.com/watch?v=31IyfM1gygo)
If I'm right about this, the issue isn't caused by adaptive-lighting or even MQTT; it's improperly configured user settings.

Thank you for sticking around to the end!

TL;DR if this issue could be written in a TL;DR shorter than this paragraph, then this issue wouldn't be 3 years old.

**Comment by th3w1zard1** [comment]
see #450

**Comment by th3w1zard1** [comment]
@broyuken This should be fixed in [1.10.0](https://github.com/basnijholt/adaptive-lighting/releases/tag/1.10.0)

**Comment by broyuken** [comment]
This isn’t fully solved, one of my lights randomly turned on at 6am this morning. The logs just show that it was turned on, not called by anything.

**Comment by th3w1zard1** [comment]
If you can post logs showing adaptive lighting fired the light.turn_on command I'll happily reopen this issue and get a priority status on this issue for the next release.

**Comment by broyuken** [comment]
That’s the problem, light.turn_on isn’t called. It just says the light was turned on.

**Comment by th3w1zard1** [comment]
@broyuken 

There's only two situations where adaptive lighting does `anything` to a light _specifically_:
1. when it calls `light.turn_on` which will be logged every time.
2. on every interval, if `detect_non_ha_changes: true`, adaptive lighting will call `async_update_entity` to force a state update for your bulb.

If you're not getting the log message then only three possibilities remain:
1. your zigbee network has too much latency/traffic and needs either more routers or an adjustment to the config, perhaps QoS of 1 instead of 0.
2. home assistant is doing things it shouldn't be, which you can replicate yourself by copying every light.turn_on adaptive lighting does from the logs into a script to test yourself.
3. `async_update_entity` is scaring your bulb into confusion somehow.

if `light.turn_on` was not called by the integration, there's nothing we can do on our end, as we're not _doing_ anything on our end if you catch my drift...

See #506 to ensure you have the proper config options set for your lights for use with this integration.

**Comment by basnijholt** [comment]
Great news, there might be a solution! :tada: 🚀 

I also had a major insight yesterday going through the code line-by-line.

I believe the issue of lights turning on is always because for some reason: _the light will incorrectly report its state as on!_ (as mentioned by @th3w1zard1 above).

Because AL then notices a state change, and this state is likely not precisely what it should be according to AL, it will adapt it to the correct state, inadvertently turning the lights on.

In this PR https://github.com/basnijholt/adaptive-lighting/pull/663, I have made some changes. With that code when `detect_non_ha_changes: false` and there is a state change which is **not** the result of a `light.turn_on`, it will mark the light as manually controlled and not adapt it at all.

I have one light at home which frequently turns on again after I turned it off, I will see whether this still happens with the changes. If not, I will merge and release a new version.

**Comment by basnijholt** [comment]
Hi folks!

I have merged https://github.com/basnijholt/adaptive-lighting/pull/663 and released [v1.18.0](https://github.com/basnijholt/adaptive-lighting/releases/tag/1.18.0). Set `detect_non_ha_changes: false` for the switches that contain the lights that accidentally turn on.

I will close this issue and would really appreciate folks reporting back on whether the issue is fixed. ⚙️ 

Thanks everyone for your support and patience in getting this issue fixed! 🙌❤️

**Comment by basnijholt** [comment]
BTW if you have debugging logging enabled you'll see messages like these:

`2023-07-28 12:40:00.233 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Bedroom: Ignoring 'off' → 'on' event for 'light.baby_room' with context.id='01H6F07PV5V2HD7W30083Y8RZS' because 'light.turn_on' was not called by HA and 'detect_non_ha_changes' is False`

This would previously have turned on the light 💡

**Comment by Dinth** [comment]
Hiya. This doesnt seem to work for me. When im changing device state via HA everything works perfectly fine, but whenever i switch off lights via a zigbee switch bound to a group of lights (Adaptive lighting is also configured for that group, not individual lights), after around 2 minutes they came back on.

**Comment by wise-rice** [comment]
I also have this issue. Lights *sometimes* turn back on a few seconds after being turned off.
There is a related issue here: #21286 (might be related to only LIDL or Tuya lights? might be an AdaptiveLighting Problem)

detect_non_ha_changes is turned off / seems broken (does not recognize changes made through light link group dimmers), anyway.I wish I could use it, though.
